# scPipeline

The [scPipeline](https://github.com/NMikolajewicz/scPipeline) analysis pipeline is a modular collection of R markdown scripts that generate [flexdashboard](https://rstudio.github.io/flexdashboard/index.html) reports for scRNAseq analyses, and is dependent on the [scMiko](https://github.com/NMikolajewicz/scMiko) package. The modular framework permits flexible usage, without necessitating complete reliance on a single analysis platform. Additionally, the self-contained reports that are generated by each module provide a comprehensive analysis summary and log of analytic parameters and scripts, thereby ensuring reproducible and shareable analysis workflows. 

To get started with scPipeline, please refer to our [scMiko](https://nmikolajewicz.github.io/scMiko/) and [scPipeline](https://nmikolajewicz.github.io/scMiko/) tutorials.   

if using scPipeline and scMiko, please consider citing our work: [Mikolajewicz, N., Gacesa, R., Aguilera-Uribe, M., Brown, K. R., Moffat, J., & Han, H. (2022). Multi-level cellular and functional annotation of single-cell transcriptomes using scPipeline. Communications Biology, 5(1), 1-14.](https://www.nature.com/articles/s42003-022-04093-2)

## Dependencies

- Make sure you have R 4.4 or newer.
  - Bioconductor 3.19 only works on R 4.4
  - Certain version of packages only exist in certain version of Bioconductor.
- Just use my `renv.lock` file.
- Make sure you have RTools installed if using Windows.

```R
renv::init()

# then select 1: Restore the project from the lockfile.
```

You can use [10X PBMC](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz) as an example.
